store/models/eth.go

```diff
@@ -102,6 +102,14 @@ func (tx *Tx) AssignTxAttempt(txat *TxAttempt) {
 	tx.SentAt = txat.SentAt
 }
 
+// IsBumpableAttempt returns true only if the attempt is the last
+// attempt associated meaning it has the highest gas price after subsequent
+// bumps, or is the attempt that has received a TxReceipt,
+// being marked by MarkTxReceipt.
+func (tx *Tx) IsBumpableAttempt(txat *TxAttempt) bool {
```

Comment:
I get the intention here, but this function makes me uncomfortable - it doesn't really do what it says it does, I can pass any transaction attempt here, and if it matches the `tx`, this will return true, regardless if it's the last transaction or not. I find `if txAttempt.Hash != latestTx.Hash {` much clearer and not error prone.

Reply:
In this PR we discovered that the latest tx attempt isn't always bumpable. A prior txattempt could have been confirmed, where we must reassign that attempt onto the parent Tx and be sure not to bump other txattempts.

Example:

```
1. unconfirmed
then
1. unconfirmed
2. created
then
1.confirmed
2. mistakenly gets bumped because we iterate over all attempts
```

Even with that change however, the `nonce is too low` issue still surfaced, so I shall continue looking. But I wanted to note that it isn't always simply the latest.

Reply:
Adding logness for completeness and posterity:

```
{"level":"debug","ts":1553187870.1497505,"caller":"store/tx_manager.go:496","msg":"Unconfirmed TX 717 attempt 0x48d74e077e604c14fca9352668aeb290d6d7517b67d2a43be32910c4866cbb76, bumping gas","txHash":"0x48d74e077e604c14fca9352668aeb290d6d7517b67d2a43be32910c4866cbb76","txid":717,"nonce":716,"gasPrice":"20000000000","from":"0x696b07b94A1045E4BfE57D9c33FebF19a1b47809","blkNum":367,"sentAt":354}
{"level":"info","ts":1553187870.1583552,"caller":"store/tx_manager.go:519","msg":"Bumping gas to 25000000000 for transaction 0x4c8b8b0b1f7833f30c77adc5e76e522854a2593527204861f9ff162df4ce15de","txat":{"Hash":"0x4c8b8b0b1f7833f30c77adc5e76e522854a2593527204861f9ff162df4ce15de","TxID":717,"GasPrice":25000000000,"Confirmed":false,"Hex":"0xf9012e8202cc8505d21dba008307a120946f72e8ddff7cdde74085e6a0a3cfdd27428f347880b8c44ab0d19071eed29ebce15083350a113db394831df00dae6b986f09397f8a09444c2042050000000000000000000000000000000000000000000000000de0b6b3a764000000000000000000000000000010e7d92d875c880b2274af71cb5c436384b2d45c92cdaaf300000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000005c93c52900000000000000000000000000000000000000000000000000000000000034ed820a96a02dcb238b87c15fa4bd8986c31da3fb57b946f4c16a59ec83b8a2a6260bdafd15a07d1c14ee46cbb3f3ed02257f9245f2560cfc7d7638acebddb9efc26a94a0aba8","SentAt":367,"CreatedAt":"2019-03-21T17:04:30.1571613Z"}}
{"level":"debug","ts":1553187876.5229065,"caller":"store/tx_manager.go:496","msg":"Unconfirmed TX 717 attempt 0x48d74e077e604c14fca9352668aeb290d6d7517b67d2a43be32910c4866cbb76, bumping gas","txHash":"0x48d74e077e604c14fca9352668aeb290d6d7517b67d2a43be32910c4866cbb76","txid":717,"nonce":716,"gasPrice":"20000000000","from":"0x696b07b94A1045E4BfE57D9c33FebF19a1b47809","blkNum":370,"sentAt":354}
{"level":"debug","ts":1553187876.5337644,"caller":"store/tx_manager.go:436","msg":"Confirmed TX: 717 attempt 0x4c8b8b0b1f7833f30c77adc5e76e522854a2593527204861f9ff162df4ce15de waiting on 12 confirmations","txHash":"0x4c8b8b0b1f7833f30c77adc5e76e522854a2593527204861f9ff162df4ce15de","txid":717,"nonce":716,"gasPrice":"25000000000","from":"0x696b07b94A1045E4BfE57D9c33FebF19a1b47809","receiptBlockNumber":"369","currentBlockNumber":370,"receiptHash":"0x4c8b8b0b1f7833f30c77adc5e76e522854a2593527204861f9ff162df4ce15de","confirmedAt":"380"}
{"level":"error","ts":1553187876.5382397,"caller":"adapters/eth_tx.go:110","msg":"EthTx Adapter Perform Resuming: bumpGas: TxManager#sendTransaction with nonce 716: nonce too low","stacktrace":"github.com/smartcontractkit/chainlink/logger.Error\n\tC:/devgo/src/github.com/smartcontractkit/chainlink/logger/logger.go:155\ngithub.com/smartcontractkit/chainlink/adapters.ensureTxRunResult\n\tC:/devgo/src/github.com/smartcontractkit/chainlink/adapters/eth_tx.go:110\ngithub.com/smartcontractkit/chainlink/adapters.(*EthTx).Perform\n\tC:/devgo/src/github.com/smartcontractkit/chainlink/adapters/eth_tx.go:45\ngithub.com/smartcontractkit/chainlink/services.executeTask\n\tC:/devgo/src/github.com/smartcontractkit/chainlink/services/job_runner.go:225\ngithub.com/smartcontractkit/chainlink/services.executeRun\n\tC:/devgo/src/github.com/smartcontractkit/chainlink/services/job_runner.go:248\ngithub.com/smartcontractkit/chainlink/services.(*jobRunner).workerLoop\n\tC:/devgo/src/github.com/smartcontractkit/chainlink/services/job_runner.go:161\ngithub.com/smartcontractkit/chainlink/services.(*jobRunner).channelForRun.func1\n\tC:/devgo/src/github.com/smartcontractkit/chainlink/services/job_runner.go:139"}
{"level":"debug","ts":1553187877.981461,"caller":"store/tx_manager.go:436","msg":"Confirmed TX: 717 attempt 0x4c8b8b0b1f7833f30c77adc5e76e522854a2593527204861f9ff162df4ce15de waiting on 12 confirmations","txHash":"0x4c8b8b0b1f7833f30c77adc5e76e522854a2593527204861f9ff162df4ce15de","txid":717,"nonce":716,"gasPrice":"25000000000","from":"0x696b07b94A1045E4BfE57D9c33FebF19a1b47809","receiptBlockNumber":"369","currentBlockNumber":371,"receiptHash":"0x4c8b8b0b1f7833f30c77adc5e76e522854a2593527204861f9ff162df4ce15de","confirmedAt":"380"}
{"level":"debug","ts":1553187880.9781659,"caller":"store/tx_manager.go:436","msg":"Confirmed TX: 717 attempt 0x4c8b8b0b1f7833f30c77adc5e76e522854a2593527204861f9ff162df4ce15de waiting on 12 confirmations","txHash":"0x4c8b8b0b1f7833f30c77adc5e76e522854a2593527204861f9ff162df4ce15de","txid":717,"nonce":716,"gasPrice":"25000000000","from":"0x696b07b94A1045E4BfE57D9c33FebF19a1b47809","receiptBlockNumber":"369","currentBlockNumber":372,"receiptHash":"0x4c8b8b0b1f7833f30c77adc5e76e522854a2593527204861f9ff162df4ce15de","confirmedAt":"380"}
{"level":"debug","ts":1553187885.1302104,"caller":"store/tx_manager.go:436","msg":"Confirmed TX: 717 attempt 0x4c8b8b0b1f7833f30c77adc5e76e522854a2593527204861f9ff162df4ce15de waiting on 12 confirmations","txHash":"0x4c8b8b0b1f7833f30c77adc5e76e522854a2593527204861f9ff162df4ce15de","txid":717,"nonce":716,"gasPrice":"25000000000","from":"0x696b07b94A1045E4BfE57D9c33FebF19a1b47809","receiptBlockNumber":"369","currentBlockNumber":375,"receiptHash":"0x4c8b8b0b1f7833f30c77adc5e76e522854a2593527204861f9ff162df4ce15de","confirmedAt":"380"}
{"level":"debug","ts":1553187890.6538033,"caller":"store/tx_manager.go:436","msg":"Confirmed TX: 717 attempt 0x4c8b8b0b1f7833f30c77adc5e76e522854a2593527204861f9ff162df4ce15de waiting on 12 confirmations","txHash":"0x4c8b8b0b1f7833f30c77adc5e76e522854a2593527204861f9ff162df4ce15de","txid":717,"nonce":716,"gasPrice":"25000000000","from":"0x696b07b94A1045E4BfE57D9c33FebF19a1b47809","receiptBlockNumber":"369","currentBlockNumber":377,"receiptHash":"0x4c8b8b0b1f7833f30c77adc5e76e522854a2593527204861f9ff162df4ce15de","confirmedAt":"380"}
{"level":"debug","ts":1553187898.9451382,"caller":"store/tx_manager.go:436","msg":"Confirmed TX: 717 attempt 0x4c8b8b0b1f7833f30c77adc5e76e522854a2593527204861f9ff162df4ce15de waiting on 12 confirmations","txHash":"0x4c8b8b0b1f7833f30c77adc5e76e522854a2593527204861f9ff162df4ce15de","txid":717,"nonce":716,"gasPrice":"25000000000","from":"0x696b07b94A1045E4BfE57D9c33FebF19a1b47809","receiptBlockNumber":"369","currentBlockNumber":381,"receiptHash":"0x4c8b8b0b1f7833f30c77adc5e76e522854a2593527204861f9ff162df4ce15de","confirmedAt":"380"}
{"level":"info","ts":1553187898.9495769,"caller":"store/tx_manager.go:462","msg":"Confirmed TX 717: 0x4c8b8b0b1f7833f30c77adc5e76e522854a2593527204861f9ff162df4ce15de","txHash":"0x4c8b8b0b1f7833f30c77adc5e76e522854a2593527204861f9ff162df4ce15de","txid":717,"nonce":716,"ethBalance":"999383677955000000000","linkBalance":"0.000000000000000000","receipt":{"blockNumber":369,"transactionHash":"0x4c8b8b0b1f7833f30c77adc5e76e522854a2593527204861f9ff162df4ce15de"},"txat":{"Hash":"0x4c8b8b0b1f7833f30c77adc5e76e522854a2593527204861f9ff162df4ce15de","TxID":717,"GasPrice":25000000000,"Confirmed":true,"Hex":"0xf9012e8202cc8505d21dba008307a120946f72e8ddff7cdde74085e6a0a3cfdd27428f347880b8c44ab0d19071eed29ebce15083350a113db394831df00dae6b986f09397f8a09444c2042050000000000000000000000000000000000000000000000000de0b6b3a764000000000000000000000000000010e7d92d875c880b2274af71cb5c436384b2d45c92cdaaf300000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000005c93c52900000000000000000000000000000000000000000000000000000000000034ed820a96a02dcb238b87c15fa4bd8986c31da3fb57b946f4c16a59ec83b8a2a6260bdafd15a07d1c14ee46cbb3f3ed02257f9245f2560cfc7d7638acebddb9efc26a94a0aba8","SentAt":370,"CreatedAt":"0001-01-01T00:00:00Z"},"err":null}

```

Reply:
I have since proven this scenario incorrect:

```
1. unconfirmed
then
1. unconfirmed
2. created
then
1.confirmed
2. mistakenly gets bumped because we iterate over all attempts
```